### PR TITLE
373 slices_restore will restore Dates and POSIXT classes

### LINKED ISSUE #373: slices_restore do not restore POSIXT and DATES classes
### What happened?

When storing a date with slices_store it gets saved as character. On a restore it remains a characte.

# Code

```{R}
x <- 
  filter_var(
    dataname = "ADSL", 
    varname = "COUNTRY", 
    selected = as.POSIXct(c("2023-06-29 18:05:32 CEST", "2023-06-29 18:15:32 CEST")), 
    fixed = TRUE
  )
class(as.list(x)$selected)
y <- filter_settings(x)
slices_store(y, file = 'test1.json')

z <- slices_restore('test1.json')

class(shiny::isolate(z[[1]]$selected))
```
# Code with results

```{R}
> x <- 
+   filter_var(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 18:05:32 CEST", "2023-06-29 18:15:32 CEST")), 
+     fixed = TRUE
+   )
> class(as.list(x)$selected)
[1] "POSIXct" "POSIXt" 
> y <- filter_settings(x)
> slices_store(y, file = 'test1.json')
> 
> z <- slices_restore('test1.json')
> 
> class(shiny::isolate(z[[1]]$selected))
[1] "character" 
```

### sessionInfo()

_No response_

### Relevant log output

_No response_

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct.

### Contribution Guidelines

- [X] I agree to follow this project's Contribution Guidelines.

### Security Policy

- [X] I agree to follow this project's Security Policy.

### PR DESCRIPTION
**Pull request by m7pr**
Closes #373 
A very small proposition on how we can guess that a character contains a Date a POSIXT time. Up to be discussed as we might want to just use jsonlite::serializeJSON and jsonlite::unserializeJSON

# The code

```{R}
x <- 
  teal_slice(
    dataname = "ADSL", 
    varname = "COUNTRY", 
    selected = as.POSIXct(c("2023-06-29 18:05:32 CEST", "2023-06-29 18:15:32 CEST")), 
    fixed = TRUE
  )
class(as.list(x)$selected)

t <- 
  teal_slice(
    dataname = "ADSL", 
    varname = "TIME", 
    selected = as.Date(c("2023-06-29", "2023-06-30")), 
    fixed = TRUE
  )
class(as.list(t)$selected)

y <- teal_slices(x, t)
slices_store(y, file = 'test1.json')

z <- slices_restore('test1.json')

class(shiny::isolate(z[[1]]$selected))
class(shiny::isolate(z[[2]]$selected))

````

# The code with results

```{R}
> x <- 
+   teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 18:05:32 CEST", "2023-06-29 18:15:32 CEST")), 
+     fixed = TRUE
+   )
> class(as.list(x)$selected)
[1] "POSIXct" "POSIXt" 
> 
> t <- 
+   teal_slice(
+     dataname = "ADSL", 
+     varname = "TIME", 
+     selected = as.Date(c("2023-06-29", "2023-06-30")), 
+     fixed = TRUE
+   )
> class(as.list(t)$selected)
[1] "Date"
> 
> y <- teal_slices(x, t)
> slices_store(y, file = 'test1.json')
> 
> z <- slices_restore('test1.json')
> 
> class(shiny::isolate(z[[1]]$selected))
[1] "POSIXct" "POSIXt" 
> class(shiny::isolate(z[[2]]$selected))
[1] "Date"
```

### CODE REVIEW

**Comment by chlebowa** [comment]
I am concerned about time zones.
```
x <- 
  teal_slice(
    dataname = "ADSL", 
    varname = "COUNTRY", 
    selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "BRA"), 
    fixed = TRUE
  )

y <- teal_slices(x)
slices_store(y, file = 'test1.json')

z <- slices_restore('test1.json')

identical(as.list(y, recursive = TRUE), as.list(z, recursive = TRUE))
[1] FALSE
```

**Comment by m7pr** [comment]
Indeed `tzone` attribute is dropped during the character conversion in store. One more addition would be the collapse of `tzone` argument at the storage phase to the actual string representation, so that it can be restored with `timezone <- substr(slice$selected[1], 21, nchar(slice$selected[1]))` and passed to `as.POSIXct(slice$selected, tz = timezone)`.

**Comment by m7pr** [comment]
checkout last commit, where I tried to unify stored POSIX to UTC. When storing we change the timezone so the display is displayed in UTC, and when restoring we then know we need to set timezone to UTC. The print/format is different after the restorage but the time representation (as number of seconds from 1970-01-01) is the same

```{R}
x <- 
  teal_slice(
    dataname = "ADSL", 
    varname = "COUNTRY", 
    selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "NZ"), 
    fixed = TRUE
  )

y <- teal_slices(x)
slices_store(y, file = 'test1.json')

z <- slices_restore('test1.json')

identical(
  as.integer(shiny::isolate(z[[1]]$selected)), 
  as.integer(shiny::isolate(x$selected))
)


```

**Comment by m7pr** [comment]
TODO: review if `teal.slice` works if we keep factors converted to characters

File: R/teal_slice-store.R [hunk 1]
```diff
@@ -51,6 +53,22 @@ slices_restore <- function(file) {
   checkmate::assert_file_exists(file, access = "r", extension = "json")
 
   tss_json <- jsonlite::fromJSON(file, simplifyDataFrame = FALSE)
+  tss_json$slices <-
+    lapply(tss_json$slices, function(slice) {
+      if (!is.null(slice$selected)) {
+        slice$selected <-
+          if (all(grepl('[0-9]{4}-[0-9]{2}-[0-9]{2}', slice$selected))) {
+            if (all(grepl('T[0-9]{2}:[0-9]{2}:[0-9]{2}', slice$selected))) {
+              as.POSIXct(gsub('T|Z', ' ', slice$selected), tz = 'UTC')
```
Comment (averissimo) [inline R/teal_slice-store.R:62]: How about restoring on the local timezone?

This also has the side effects of allowing for the addition of tests. See branch [373_add_tests](https://github.com/insightsengineering/teal.slice/compare/373_restore_classes%40main...373_add_tests%40main?expand=1) that merges to this one

```suggestion
              local_time <- as.POSIXct(gsub("T|Z", " ", slice$selected), tz = "UTC")
              attr(local_time, "tzone") <- Sys.timezone()
              local_time
```

File: R/teal_slice-store.R [hunk 2]
```diff
@@ -51,6 +53,22 @@ slices_restore <- function(file) {
   checkmate::assert_file_exists(file, access = "r", extension = "json")
 
   tss_json <- jsonlite::fromJSON(file, simplifyDataFrame = FALSE)
+  tss_json$slices <-
+    lapply(tss_json$slices, function(slice) {
+      if (!is.null(slice$selected)) {
+        slice$selected <-
```
Comment (averissimo) [inline R/teal_slice-store.R:59]: How about the `choices`? I think we may want to restore it as it can take `Date` or `POSIXct`

The test below fails

```R
test_that("teal_slice store/restore supports saving `Date` timestamps in choices", {
  slices_path <- withr::local_file("slices.json")

  time_stamps <- Sys.Date() + c(-10 * 600, -30, 0)

  tss <- teal_slices(
    teal_slice(
      dataname = "ADSL",
      varname = "COUNTRY",
      selected = sample(time_stamps, 2),
      choices = time_stamps,
      fixed = TRUE
    )
  )

  # Store the teal_slices object to a file
  slices_store(tss, slices_path)
  tss_restored <- slices_restore(slices_path)

  tss_restored_list <- shiny::isolate(shiny::reactiveValuesToList(tss_restored[[1]]))
  expect_s3_class(tss_restored_list$choices, "Date")

  expect_identical_slice(tss[[1]], tss_restored[[1]])
})
```

File: R/teal_slice-store.R (hunk 1 above)
Comment (averissimo) [inline R/teal_slice-store.R:62]: Styler changes :roll_eyes: 

```suggestion
          if (all(grepl("[0-9]{4}-[0-9]{2}-[0-9]{2}", slice$selected))) {
            if (all(grepl("T[0-9]{2}:[0-9]{2}:[0-9]{2}", slice$selected))) {
              as.POSIXct(gsub("T|Z", " ", slice$selected), tz = "UTC")
```

**Review by averissimo — COMMENTED** [review]
Great job on this feature!! :100: 

I thought on adding some tests. See branch [373_add_tests](https://github.com/insightsengineering/teal.slice/compare/373_restore_classes%40main...373_add_tests%40main?expand=1) that merges to this one.

Note: it skips the last 2 tests for now and might be enabled/removed depending on feedback

File: R/teal_slice-store.R (hunk 2 above)
Comment (m7pr) [reply R/teal_slice-store.R:59]: Thanks for the input @averissimo - valid point

Comment (m7pr) [reply R/teal_slice-store.R]: Thanks, just extended in here d7aebc9cac60ac4979719ef40888fdbd0d2d4b13

**Comment by m7pr** [comment]
Thanks for the branch with new tests. Just merged to this branch. Still need to figure out what happens for factors

**Comment by m7pr** [comment]
Hey @averissimo and @chlebowa the issue for `POSIXct` and `Date` classes are fixed with this PR (this closes https://github.com/insightsengineering/teal.slice/issues/373). 

Similar issue appears for `factors` that are not restored properly, but I created a separate issue for that https://github.com/insightsengineering/teal.slice/issues/467 as this is a bigger deal and became out of scope of this quick change

**Review by averissimo — COMMENTED** [review]
It looks good to me :100:

I won't approve right now to wait for @chlebowa feedback as he commented before

I was a bit concern about using `all(grepl(...`, but after some tests I can see that `teal_slices` doesn't work well with a mix bag of `selected`

```R
time_stamps <- Sys.time() + c(-10 * 60 * 60 * 24, -30, 0)
time_stamps <- as.POSIXct(ceiling(as.double(time_stamps)), tz = "UTC")

# Converts Date to POSIXct
teal_slices(
  teal_slice(
    dataname = "ADSL",
    varname = "EOSDTM",
    selected = c(time_stamps, Sys.Date()),
    fixed = TRUE
  )
)

# Converts POSIXct to Date
teal_slices(
  teal_slice(
    dataname = "ADSL",
    varname = "EOSDTM",
    selected = c(Sys.Date(), time_stamps),
    fixed = TRUE
  )
)

# Assumes all as strings
tss <- teal_slices(
  teal_slice(
    dataname = "ADSL",
    varname = "EOSDTM",
    selected = c("A string", time_stamps, Sys.Date()),
    fixed = TRUE
  )
)

# Fails to convert string to Posix.ct (acceptable)
tss <- teal_slices(
  teal_slice(
    dataname = "ADSL",
    varname = "EOSDTM",
    selected = c(time_stamps, Sys.Date(), "A string"),
    fixed = TRUE
  )
)
```

**Comment by m7pr** [comment]
@averissimo please bare in mind, that we will store and restore data provided by end app users and app developers and for those it is hard to mix such structure. End app users will have shiny inputs to provided the data, and the app developer will rather start with direct names of levels or even pull names directly from data. I think mixing classes is unlikely to happen

**Comment by chlebowa** [comment]
@averissimo The coercions you listed are done by `c`, actually.

**Comment by averissimo** [comment]
@chlebowa & @m7pr I'm always a bit pessimistic with weird edge cases :sweat_smile:

The only semi-realistic scenario would be if the user deliberately wants to break the application.

```R
tss <- teal_slices(
  teal_slice(
    dataname = "ADSL",
    varname = "EOSDTM",
    selected = c(
      "beta 2023-09-11",
      "release candidate 2023-09-21",
      "release 2023-09-21"
    ),
    fixed = TRUE
  )
)

slices_store(tss, slices_path)
tss_restored <- slices_restore(slices_path)
```

**Comment by chlebowa** [comment]
While I am not a fan of the regex approach used here, I have yet to figure out something better.

> @chlebowa & @m7pr I'm always a bit pessimistic with weird edge cases 😅

Good! :+1: 

> The only semi-realistic scenario would be if the user deliberately wants to break the application.

Should we care for those? Honest question.

> ```r
> tss <- teal_slices(
>   teal_slice(
>     dataname = "ADSL",
>     varname = "EOSDTM",
>     selected = c(
>       "beta 2023-09-11",
>       "release candidate 2023-09-21",
>       "release 2023-09-21"
>     ),
>     fixed = TRUE
>   )
> )
> 
> slices_store(tss, slices_path)
> tss_restored <- slices_restore(slices_path)
> # Error in charToDate(x) :
> # character string is not in a standard unambiguous format
> ```

Well, this brings up an interesting question. We did not consider someone storing and restoring a manually typed `teal_slices`, rather ones spit out by an application. Granted, in some cases they would be exactly as typed out by the app dev but there are checks they have to pass when the app is run. If you use that `teal_slice` with `init_filter_state` where `x` is `Date`s, none of those values would be selected.

**Comment by m7pr** [comment]
Hey @averissimo thanks for bringing this up. I think this was very important to get such edge cases validated. I did update the extraction function and extended tests with this example.

@chlebowa when it comes to regular expression usage, this is not that uncommon in IT world to specify the format and the schema of expected results and then use regular expressions to validate the format. Regular expressions are used to validate phone address, emails and dates in various forms over the web. We use the same approach to distinguish dates and POSIX*t from regular strings. That's it. I don't think we should come up with anything more for the restoring.

Any chance you can have a look again?

**Review by averissimo — APPROVED** [review]
I was unsure whether there would be any real benefits from the added complexity on the regex.

Thinking a bit more it makes sense as it needs a full match and there won't be any downsides.

I'm glad you've included it. Looks good!

**Comment by m7pr** [comment]
any last words @chlebowa ?

**Comment by chlebowa** [comment]
Let me complain some more.

```
> x <- 
+   teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "NZ"), 
+     fixed = TRUE
+   )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-28T12:00:01Z", "2023-06-28T12:00:02Z"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
```

Are we ok with this print format? `"selected" : ["2023-06-28T12:00:01Z", "2023-06-28T12:00:02Z"],`

**Comment by m7pr** [comment]
hmmm @chlebowa hmmm
For print, we could just go with `as.character` or append the end of the line with a timezone comment
so it's either
```r
"selected" : ["2023-06-29 00:00:01", "2023-06-29 00:00:02"],
```
or

```r
"selected" : ["2023-06-28T12:00:01Z", "2023-06-28T12:00:02Z"], # timezone = UTC+0
```

**Comment by chlebowa** [comment]
How? Yet another argument to `format`?

**Comment by m7pr** [comment]
@chlebowa based on the class of the `slices$selected` and `slices$choices` field.

Or!, even better, we can print a message that timestamps are printed in UTC+0 timestamp, if one of `slices$selected` and `slices$choices` fields is of POSIX class?

**Comment by chlebowa** [comment]
So when would that happen? In `print`?

**Comment by m7pr** [comment]
Yep, print would add an extra statement/note on the top or at the bottom of the print. How does that sound?

**Comment by chlebowa** [comment]
> Yep, print would add an extra statement/note on the top or at the bottom of the print. How does that sound?

I am not entirely happy to do this.



> or we extend the print and the format (hence store) so that they append new field called `selected_timezone` or `choices_timezone` with the timezone specs, depending on the class of `selected` or `choices` fields

I definitely don't like this. The print should describe the filter state, not the minutiae of a `teal_slice`.

**Comment by m7pr** [comment]
what about this? https://github.com/insightsengineering/teal.slice/pull/432#issuecomment-1744830760
which would happen based on the `selected/choices` class

**Comment by m7pr** [comment]
I think both print and format (R console print and stored file) can have the same format of having `# timezonte UTC+0` comment at the end of selected/choices if they are POSIX. It will be the same in both places

**Comment by m7pr** [comment]
hey @gogonzo any thoughts on this?

**Review by gogonzo — APPROVED** [review]
Looks good 👍

**Comment by m7pr** [comment]
Hey @gogonzo and @chlebowa - thanks for the highly productive and fruitful discussion today.
In the end we decided that `POSIX*t` classes will be stored in their `format(usetz  = TRUE)` representation so that timezone is stored next to a timestamp in a visible way, in stored file and in the print in console.

The final proposition of the usage looks like below:

```r
> x <- teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "EST"), 
+     fixed = TRUE
+ )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 EST", "2023-06-29 00:00:02 EST"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> slices_store(y, file = 'store.json')
> z <- slices_restore('store.json')
> z
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 EST", "2023-06-29 00:00:02 EST"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
```

# TODO

One thing to be fixed (ON TODO LIST) is that some timezones are named differently in `format()` and in `as.POSIXct()`.

```r
x <- teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "NZ"), 
+     fixed = TRUE
+ )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 NZST", "2023-06-29 00:00:02 NZST"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> slices_store(y, file = 'store.json')
> z <- slices_restore('store.json')
Warning messages:
1: In strptime(xx, f, tz = tz) : unknown timezone 'NZST'
2: In as.POSIXct.POSIXlt(x) : unknown timezone 'NZST'
3: In strptime(x, f, tz = tz) : unknown timezone 'NZST'
4: In as.POSIXct.POSIXlt(as.POSIXlt(x, tz, ...), tz, ...) :
  unknown timezone 'NZST'
> z
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 GMT", "2023-06-29 00:00:02 GMT"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
Warning message:
In as.POSIXlt.POSIXct(x, tz) : unknown timezone 'NZST'
```

the same happens for CET that gets changed to CEST, where CEST is not recognized by `as.POSIXct`

Will work on that and will keep you posted.

**Comment by chlebowa** [comment]
The validity of time zone codes is platform-dependent, I'm afraid.

**Comment by m7pr** [comment]
lemme investigate :)

**Comment by averissimo** [comment]
:angry: this is an upstream bug from `jsonlite` as the [ISO8601](https://en.wikipedia.org/wiki/ISO_8601#Time_zone_designators) standard should ensure that only valid timezones are used (none, UTC or UTC offset)

**Comment by chlebowa** [comment]
Ah, but we are stepping away from ISO8601 and trying to convert to strings by ourselves and store those.

**Comment by averissimo** [comment]
Oh my bad then. I missed it in the comments as the code doesn't seem to apply that

If you're considering using the `format(..., usetz = TRUE)` function explicitly, then having a manual format param might be an option to fully control the process _(with `%z` - lowercase - to deal with timezones)_:

```R
format(Sys.time(), format = "%Y-%m-%d %H:%M:%S %z", usetz = FALSE)
as.POSIXct(c("2023-06-29 17:02:01 +0200"), tryFormats = "%Y-%m-%d %H:%M:%S %z")
```

**Comment by m7pr** [comment]
This doesn't appear if we change the class of final restored object from `POSIXct` to `POSIXlt`. Assuming we know a timezone exists in the stored file, should we use `POSIXlt` ? https://github.com/insightsengineering/teal.slice/pull/432/commits/cd218bf66ba1e96d45c21d863a5ac547458dff20

```r
x <- teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "CET"), 
+     fixed = TRUE
+ )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 CEST", "2023-06-29 00:00:02 CEST"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> slices_store(y, file = 'store.json')
> z <- slices_restore('store.json')
> z
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 CEST", "2023-06-29 00:00:02 CEST"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> class(shiny::isolate(z[[1]]$selected))
[1] "POSIXlt" "POSIXt"
```

**Comment by chlebowa** [comment]
> This doesn't appear if we change the class of final restored object from `POSIXct` to `POSIXlt`. Assuming we know a timezone exists in the stored file, should we use `POSIXlt` ?

:thinking: 

I would rigorously test if `POSIXlt` will properly set state with `FilterState$set_state`.

**Comment by m7pr** [comment]
Will test it out today and let you know.
We can always do as.posixct(as.posixlt()) :) in restore

**Comment by m7pr** [comment]
```r
> as.POSIXlt("2023-10-08 08:46:40 UTC", format = "%Y-%m-%d %H:%M:%S")
[1] "2023-10-08 08:46:40 CEST"
> as.POSIXlt("2023-10-08 08:46:40 CEST", format = "%Y-%m-%d %H:%M:%S")
[1] "2023-10-08 08:46:40 CEST"
```

`as.POSIXlt` on `"%Y-%m-%d %H:%M:%S %z"` does not pay attention to `%z`. Even though `%z` is discouraged in favour of `tz` parameter, the `tz` parameter has this disadvantage that it operates on short-codes that tend to be system-specific and daylight-saving-time specific. I rather go with `%z` and @averissimo proposition 

```
format(Sys.time(), format = "%Y-%m-%d %H:%M:%S %z", usetz = FALSE)
[1] "2023-10-18 10:59:27 +0200"
```

To have explicit amount of hour displayed after the timestamp. So now we only deal with numbers (and not characters that tend to change. `POSIXct` changes `CET` to `CEST` and then doesn`t understand what CEST is

```r
> format(as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "CET"), usetz = TRUE)
[1] "2023-06-29 00:00:01 CEST" "2023-06-29 00:00:02 CEST"
> as.POSIXct(format(as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "CET"), usetz = TRUE), tz = 'CEST')
[1] "2023-06-29 00:00:01 GMT" "2023-06-29 00:00:02 GMT"
Warning messages:
1: In strptime(xx, f, tz = tz) : unknown timezone 'CEST'
2: In as.POSIXct.POSIXlt(x) : unknown timezone 'CEST'
3: In strptime(x, f, tz = tz) : unknown timezone 'CEST'
4: In as.POSIXct.POSIXlt(as.POSIXlt(x, tz, ...), tz, ...) :
  unknown timezone 'CEST'
5: In as.POSIXlt.POSIXct(x, tz) : unknown timezone 'CEST'
```

**Comment by chlebowa** [comment]
See, I don't get those warnings :thinking:

**Comment by m7pr** [comment]
I just pushed @averissimo proposition, but the restore restores the timestamp into the current timezone of the session, which is not ideal. I will need to revert it and work on short-codes of timezones

```r
x <- teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "UTC"), 
+     fixed = TRUE
+ )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 00:00:01 +0000", "2023-06-29 00:00:02 +0000"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> slices_store(y, file = 'store.json')
> z <- slices_restore('store.json')
> z
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-29 02:00:01 +0200", "2023-06-29 02:00:02 +0200"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> class(shiny::isolate(z[[1]]$selected))
[1] "POSIXct" "POSIXt"
```

**Comment by m7pr** [comment]
I wonder if `lubridate` package could handle that..

**Comment by averissimo** [comment]
:+1: lubridate.. There's a reason why the meme about never working with timezones, it's full of problems.

Is it a big problem to restore the timezone into the local one?

**Comment by m7pr** [comment]
It's not an issue at all for filtering. It's only a cosmetic visual
difference in print.

**Comment by m7pr** [comment]
Ok guys, let's end this painful process. I know you are all sick and tired of timestamps already!

Based on all edge-cases that we've noticed during this discussion, I assume there is no gold solution in here. I just made final changes and the one that I propose, at least in my opinion, is the least evil.

Even though a timestamp can be passed in any timezone, we do convert it to `UTC` for print/format but we append `UTC` to print/format so that this change is visible. Under the hood the timezone is a number so filtering capabilities remain the same, no matter in which timezone we print the number.

Having UTC in print/format (hence store which uses format) allows us to restore `POSIX*t` in `UTC` as well, so they are printed in `UTC` again. 


```r
> x <- teal_slice(
+     dataname = "ADSL", 
+     varname = "COUNTRY", 
+     selected = as.POSIXct(c("2023-06-29 00:00:01", "2023-06-29 00:00:02"), tz = "CET"), 
+     fixed = TRUE
+ )
> y <- teal_slices(x)
> y
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-28 22:00:01 UTC", "2023-06-28 22:00:02 UTC"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> slices_store(y, file = 'store.json')
> z <- slices_restore('store.json')
> z
{
  "slices": [
    {
      "dataname" : "ADSL",
      "varname"  : "COUNTRY",
      "id"       : "ADSL COUNTRY",
      "selected" : ["2023-06-28 22:00:01 UTC", "2023-06-28 22:00:02 UTC"],
      "fixed"    : true,
      "anchored" : false,
      "multiple" : true
    }
  ],
  "attributes": {
    "allow_add"  : true
  }
} 
> class(shiny::isolate(z[[1]]$selected))
[1] "POSIXct" "POSIXt" 
```

**Comment by chlebowa** [comment]
Seems so be working just fine and I see that checks have passed, but some test fail for me locally :thinking:

**Comment by averissimo** [comment]
I guess the tests should use `1970-01-01`

> `r> ?as.POSIXct.numeric`
> ...
> Optional since R 4.3.0, where the equivalent of "1970-01-01" is used.

**Comment by chlebowa** [comment]
> > Optional since R 4.3.0, where the equivalent of "1970-01-01" is used.

That explains it :unamused:

**Comment by m7pr** [comment]
appended tests. thanks for local check. I think we should extend GITHUB R CMD CHECK with multiple R versions (and preferably multiple platforms)

**Comment by m7pr** [comment]
Raised a request to start running GHA Checks with multiple R versions https://github.com/insightsengineering/idr-tasks/issues/682
